File: leetCode/509斐波那契数.cpp
//#include<iostream>
//#include<vector>
//#include<queue>
//using namespace std;
//
//struct TreeNode {
//	int val;
//	TreeNode* left;
//	TreeNode* right;
//	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
//};
//
//int fib(int N) {
//	if (N < 1) {
//		return 0;
//	}
//	if (N == 1 || N == 2) {
//		return 1;
//	}
//	int curr, pre ;
//	pre = curr = 1;
//	int sum = 1;
//	for (int i = 3; i <= N; i++) {
//		sum = pre + curr;
//		pre = curr;
//		curr = sum;
//	}
//	return sum;
//}
//
//vector<vector<int>> levelOrder(TreeNode* root) {
//	vector<vector<int>>result;
//	queue<TreeNode*>queue;
//	if (root == NULL) {
//		return result;
//	}
//	queue.push(root);
//	while (!queue.empty())
//	{
//		int len = queue.size();
//		vector<int>level;
//		for (int i = 0; i < len; i++) {
//			TreeNode* node = queue.front();
//			queue.pop();
//			level.push_back(node->val);
//			if (node->left!= NULL) {
//				queue.push(node->left);
//			}
//			if (node->right != NULL) {
//				queue.push(node->right);
//			}
//		}
//		result.push_back(level);
//	}
//	return result;
//}
//
//int main() {
//	int n;
//	scanf("%d", &n);
//}

File: src/leetCode/509. 斐波那契数.js
/**
 * @param {number} n
 * @return {number}
 */
var fib = function (n) {
  let dp = Array(31).fill(0);
  dp[1] = 1;
  dp[2] = 1;
  for (let i = 3; i <= n; i++) {
    dp[i] = dp[i - 1] + dp[i - 2];
  }
  return n;
};

/**
 * @param {number} n
 * @return {number}
 */

const map = {};
var fib = function (n) {
  if (n === 1) return 1;
  if (n === 2) return 2;
  if (map[n]) {
    return map[n];
  }
  let result = fib(n - 1) + fib(n - 2);
  map[n] = result;
  return result;
};
